## Jacknife

In [1]:
import numpy as np

In [2]:
def Jacknife(X , estimador, q=0.75):

#######################################################

    if estimador == 'median':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.median( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.median(X) )

        estimacion_sesgo_corregido = np.median(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

 #######################################################   

    if estimador == 'mean':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.mean( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.mean(X) )

        estimacion_sesgo_corregido = np.mean(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'std':

        def Jacknife_sample(Xj , r):

            X_sample_r = np.delete(Xj, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.std( Jacknife_sample(X , r) ) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.std(X) )

        estimacion_sesgo_corregido = np.std(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'quantile':

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( np.quantile( Jacknife_sample(X , r) , q) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.quantile(X , q) )

        estimacion_sesgo_corregido = np.quantile(X , q) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

#######################################################

    if estimador == 'kurtosis':

        import scipy

        from scipy.stats import  kurtosis

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( kurtosis(X) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - kurtosis(X) )

        estimacion_sesgo_corregido = kurtosis(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
     
#######################################################

    if estimador == 'skewness':

        import scipy

        from scipy.stats import  skew

        def Jacknife_sample(X , r):

            X_sample_r = np.delete(X, r)

            return(X_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(X)):

            replicas_estimador.append( skew(X) )

    
        n = len(X)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - skew(X) )

        estimacion_sesgo_corregido = skew(X) - sesgo

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
    
    return(sesgo, estimacion_sesgo_corregido, standard_error)

In [3]:
np.random.seed(123)

X = np.random.normal(loc=10, scale=15, size=50)

In [4]:
np.median(X)

8.23916733155832

In [5]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'median', q=0.75)

In [6]:
sesgo

8.704148513061227e-14

In [7]:
estimacion_sesgo_corregido

8.239167331558233

In [8]:
standard_error

2.381386940718188

In [9]:
np.mean(X)

10.199071616256777

In [10]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'mean')

In [11]:
sesgo

-8.704148513061227e-14

In [12]:
estimacion_sesgo_corregido

10.199071616256864

In [13]:
standard_error

2.5491917443460235

In [14]:
np.std(X)

17.844342210422187

In [15]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'std')

In [16]:
sesgo

-0.26116994703598806

In [17]:
estimacion_sesgo_corregido

18.105512157458175

In [18]:
standard_error

1.6795955569730596

In [20]:
np.quantile(X , q=0.75)

sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'quantile', q=0.75)

In [21]:
import scipy

from scipy.stats import  kurtosis

In [22]:
kurtosis(X)

sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'kurtosis')

In [23]:
sesgo

0.0

In [24]:
estimacion_sesgo_corregido

-0.37420768292897266

In [25]:
standard_error

0.0

In [26]:
import scipy

from scipy.stats import  skew

In [29]:
skew(X)


0.025587358812510053

In [30]:

sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(X , 'asimetria')

UnboundLocalError: local variable 'sesgo' referenced before assignment

In [ ]:
sesgo

In [ ]:
estimacion_sesgo_corregido

In [ ]:
standard_error

# Bootstrap

In [31]:
def Bootstrap_sample(X, random_state):

    from sklearn.utils import resample

    sample = resample( X, n_samples=len(X) , replace=True, random_state=random_state)

    return sample

In [32]:
Bootstrap_sample(X, random_state=123)

array([  9.82254258,  14.24467747,   7.8989692 ,  23.91193648,
        -3.19304515,  42.80179133,  15.79279599,   4.13650309,
        -4.03750803,  -0.49815852, -16.57299657,  43.57215008,
        16.1936824 ,  -3.00110603, -16.57299657,  45.88547899,
       -16.57299657,  16.1936824 ,   0.43372746,  15.79279599,
         3.34027061,  10.04268874, -16.57299657,  43.08895124,
         1.32099622,  43.57215008, -12.59442071,  14.24467747,
        21.06052864,  14.25440986,  14.24467747,  21.06052864,
        16.1936824 ,  24.68104009,   3.56631057, -15.91504241,
         7.39546476,   7.8989692 ,  -3.19304515,  -0.49815852,
        32.36098042,   6.16570944, -11.4302105 ,  23.91193648,
        -0.49815852,  32.37084439,  -2.08049777, -12.59442071,
         4.13650309,  34.77154806])

In [38]:
def Bootstrap(X , B, estimador, q=0.75, random_seed=123 ):

#######################################################

    np.random.seed(random_seed)

#######################################################

    if estimador == 'median':

        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X))

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( np.median( Bootstrap_sample(X) ) )

    
        sesgo = np.mean( replicas_estimador ) - np.median(X) 

        estimacion_sesgo_corregido = np.median(X) - sesgo

        standard_error = np.std( replicas_estimador ) 

#######################################################

    if estimador == 'mean':

        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X) , replace=True)

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( np.mean( Bootstrap_sample(X) ) )

    
        sesgo = np.mean( replicas_estimador ) - np.mean(X) 

        estimacion_sesgo_corregido = np.mean(X) - sesgo

        standard_error = np.std( replicas_estimador ) 

#######################################################

    if estimador == 'std':

        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X) , replace=True)

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( np.std( Bootstrap_sample(X) ) )

        
        sesgo = np.mean( replicas_estimador ) - np.std(X) 

        estimacion_sesgo_corregido = np.std(X) - sesgo

        standard_error = np.std( replicas_estimador ) 

#######################################################

    if estimador == 'quantile':

        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X) , replace=True)

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( np.quantile( Bootstrap_sample(X) , q ) )

    
        n = len(X)

        sesgo = np.mean( replicas_estimador ) - np.quantile(X , q) 

        estimacion_sesgo_corregido = np.quantile(X , q) - sesgo

        standard_error = np.std( replicas_estimador ) 

#######################################################

    if estimador == 'kurtosis':

        import scipy

        from scipy.stats import kurtosis

        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X) , replace=True)

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( kurtosis( Bootstrap_sample(X) , q ) )

    
        n = len(X)

        sesgo = np.mean( replicas_estimador ) - kurtosis(X) 

        estimacion_sesgo_corregido = kurtosis(X) - sesgo

        standard_error = np.std( replicas_estimador ) 

#######################################################

    if estimador == 'skewness':

        import scipy

        from scipy.stats import skew


        def Bootstrap_sample(X):

            from sklearn.utils import resample

            sample = resample( X, n_samples=len(X) , replace=True)

            return sample

        
        replicas_estimador = []

        for b in range(0, B):

            replicas_estimador.append( skew( Bootstrap_sample(X) ) )

    
        n = len(X)

        sesgo = np.mean( replicas_estimador ) - skew(X) 

        estimacion_sesgo_corregido = skew(X) - sesgo

        standard_error = np.std( replicas_estimador ) 


#######################################################

    return sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador

In [ ]:
np.median(X)

8.23916733155832

In [46]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimador='median', random_seed=222)

In [47]:
replicas_estimador[0:5]

[8.860755889993856,
 7.032339319605527,
 10.042688738452165,
 6.165709442041046,
 8.579365465946633]

In [ ]:
sesgo

0.45027242884926366

In [ ]:
estimacion_sesgo_corregido

7.788894902709057

In [ ]:
standard_error

1.943287160277303

In [ ]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimador='mean')

In [ ]:
np.mean(X)

10.199071616256777

In [ ]:
sesgo

0.01870102960420006

In [ ]:
estimacion_sesgo_corregido

10.180370586652577

In [ ]:
standard_error

1.5955139969727374

In [ ]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimador='std')

In [ ]:
np.std(X)

17.844342210422187

In [ ]:
replicas_estimador[0:10]

In [ ]:
sesgo

-0.24366444292487444

In [ ]:
estimacion_sesgo_corregido

18.088006653347062

In [ ]:
standard_error

1.6137690836872287

In [ ]:
sesgo , estimacion_sesgo_corregido , standard_error, replicas_estimador = Bootstrap(X , B=20000, estimador='quantile', q=0.75)

In [ ]:
np.quantile(X, q=0.75)

23.835596841535178

In [ ]:
replicas_estimador[0:5]

[23.606577937004516,
 24.960181698753786,
 28.98904388058301,
 23.19908451689496,
 24.965866373402413]

In [ ]:
sesgo

-1.0135682894025813

In [ ]:
estimacion_sesgo_corregido

11.212639905659358

In [ ]:
standard_error

3.4892420733902

## Intervalos de confianza cuantil-bootstrap

### Intervalo para un parametro (media, varianza, mediana, cuantiles, asimetria, curtosis)

In [76]:
def cuantil_boot_interval(Variable, alpha, estimator , B, random_seed=123 , q=0.75):

######################################################################

    np.random.seed(random_seed)

######################################################################

    def Bootstrap_sample(Variable):

        from sklearn.utils import resample

        sample = resample( Variable, n_samples=len(Variable))

        return sample

######################################################################

    replicas_estimador = []

    if estimator == 'mean':

        for b in range(0, B):

            replicas_estimador.append( np.mean( Bootstrap_sample(Variable) ) )

        estimation = np.mean(Variable)

    
    if estimator == 'median':

        for b in range(0, B):

            replicas_estimador.append( np.median( Bootstrap_sample(Variable) ) )

        estimation = np.median(Variable)


    if estimator == 'std':

        for b in range(0, B):

            replicas_estimador.append( np.std( Bootstrap_sample(Variable) ) )

        estimation = np.std(Variable)


    if estimator == 'skewness':

        from scipy.stats import skew

        for b in range(0, B):

            replicas_estimador.append( skew( Bootstrap_sample(Variable) ) )

        estimation = skew(Variable)


    if estimator == 'kurtosis':

        from scipy.stats import kurtosis

        for b in range(0, B):

            replicas_estimador.append( kurtosis( Bootstrap_sample(Variable) ) )

        estimation = kurtosis(Variable)


    if estimator == 'quantile':

        for b in range(0, B):

            replicas_estimador.append( np.quatile( Bootstrap_sample(Variable) , q=q ) )

        estimation = np.quantile(Variable , q=q)

######################################################################

    L1 = np.quantile( replicas_estimador  , q=alpha/2)

    L2 = np.quantile( replicas_estimador  , q=1-alpha/2)

######################################################################

    interval = [L1,L2]
    
######################################################################

    return interval , estimation

### Para la media

In [69]:
interval , estimation = cuantil_boot_interval(Variable=X, alpha=0.05, estimator='mean' , B=20000, random_seed=123)

In [70]:
interval

[5.274553626813773, 15.146375204290525]

In [71]:
estimation

10.199071616256777

In [61]:
def CI_Mean(Variable , alpha=0.05):

    n = len(Variable)

    t_alpha_medios = scipy.stats.t.ppf( 1 - alpha/2 , df=n-1)

    X_mean = Variable.mean()

    X_cuasi_var = Variable.std()**2 
    
    # std() esta definida por defecto como la cuasi-desviacion-tipica

    L1 = X_mean - t_alpha_medios * np.sqrt(X_cuasi_var/n)

    L2 = X_mean + t_alpha_medios * np.sqrt(X_cuasi_var/n)

    interval = [L1 , L2]

    return interval , X_mean

In [62]:
interval , X_mean = CI_Mean(X , alpha=0.05)

In [63]:
interval

[5.127765678327374, 15.27037755418618]

### Para la desviacion tipica

In [77]:
interval , estimation = cuantil_boot_interval(Variable=X, alpha=0.05, estimator='std' , B=20000, random_seed=123)

In [78]:
interval

[14.427214668666169, 20.756590926218593]

In [79]:
estimation

17.844342210422187

In [92]:
def CI_Variance(Variable , alpha=0.05):

    n = len(Variable)

    chi_alpha_medios = scipy.stats.chi2.ppf( 1 - alpha/2 , df=n-1)

    chi_1_alpha_medios = scipy.stats.chi2.ppf(alpha/2 , df=n-1)

    X_cuasi_var = Variable.std()**2 

    X_var = ( (n-1)/n )*X_cuasi_var
    
    # std() esta definida por defecto como la cuasi-desviacion-tipica

    L1 =  (n*X_var) / chi_alpha_medios

    L2 = (n*X_var) / chi_1_alpha_medios

    interval = [L1 , L2]

    return interval , X_var 

In [95]:
interval , X_var = CI_Variance(Variable=X , alpha=0.05)

In [96]:
interval

[222.18841684285198, 494.45882436956106]

In [97]:
X_var

312.0521379442019

### Para la mediana

In [98]:
interval , estimation = cuantil_boot_interval(Variable=X, alpha=0.05, estimator='median' , B=20000, random_seed=123)

In [99]:
interval

[2.4028635430970975, 15.99323919612726]

In [100]:
estimation

8.23916733155832

### Para la asimetria

### Para la curtosis

### Para los cuantiles

## Intervalo para la diferencia de parametros (medias, medianas, varianzas, cuantiles, curtosos, asimetrias)

## Contrastes de hipotesis por permutaciones

El test de permutaciones es un test de significancia estadística para el estudio de diferencias entre grupos. Fue desarrollado por Ronald Fisher y E.J.G. Pitman en 1930. La distribución del estadístico estudiado (media, mediana...) se obtiene calculando el valor de dicho estadístico para todas las posibles reorganizaciones de las observaciones en los distintos grupos. Dado que implica calcular todas las posibles situaciones, se trata de un test exacto.



La filosofia de los test de permutaciones es la siguiente:



Se tiene un conjunto de individuos, los cuales se distribuyen aleatoriamente en dos grupos, $T$ y $C$, de tamaños  $n_t$  y  $n_c$



Se aplica un experimento, que generalmente consiste en aplicar un tratamiento sobre un grupo (grupo tratamiento $T$) y no aplicarlo sobre el otro grupo (grupo control $C$).



Se mide una variable de interés $X$ en cada uno de los grupos:

$$ X^t = (x^t_{1},...,x^t_{n_t }) $$
$$ X^c = (x^c_{1},...,x^c_{n_c }) $$

Donde: 

$x^t_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $T$

$x^c_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $C$




Se calcula un estadistico $\widehat{\theta}$ de $X$ en los grupos $\Rightarrow \widehat{\theta}_t = f(X^t), \widehat{\theta}_c = f(X^c)$

Estadisticos habituales son la media y la mediana.



Se desea  determinar si existen diferencias significativas entre el valor del estadistico para cada grupo. 

Para ello se considera que si la diferencia observada en valor absoluto ($\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$ ) es 
mayor que la diferencia esperada por la asignación aleatoria de los grupos, en valor absoluto.
 

En términos estadísticos, la hipótesis nula ( $H_0$ ) es que ambas muestras pertenecen a la misma distribución y que, las diferencias observadas, son debidas únicamente a variaciones causadas por el reparto aleatorio en grupos. 



El test de permutación permite identificar si hay evidencias en contra de esta hipótesis. Para ello los pasos a seguir son:

1) Calcular la diferencia observada del estadistico para los dos grupos  

 $$\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$$



2) Todas las observaciones se unen sin tener en cuenta el grupo al que pertenecen. $(X^t , X^c)$



3) Se generan $B$ pares de muestras  sin reemplazamiento  de $(X^t , X^c)$ , de tamaño $n_t$ y $n_c$ respectivamente :

$$(\mathcal{X}^t_1 , \mathcal{X}^c_1), (\mathcal{X}^t_2 , \mathcal{X}^c_2),\dots, (\mathcal{X}^t_B , \mathcal{X}^c_B)$$


   $\mathcal{X}^g_i$ es la muestra aleatoria $i$-esima de tamaño n_g extraida de  $(X^t , X^c)$ , para $g=t,c$ , $i=1,...,B$
   
  



4) Para cada combinación, calcular la diferencia entre medias  (difcalculada) . El conjunto de valores obtenidos forman la distribución exacta de las posibles diferencias siendo cierta la hipótesis nula de que los dos grupos proceden de la misma población. A esta distribución, se le conoce como permutation distribution of the mean difference.

5) Calcular el p-value de dos colas como la proporción de combinaciones en las que, el valor absoluto de la diferencia calculada, es mayor o igual al valor absoluto de la diferencia observada. La idea es que si esta proporcion no es suficientemente pequeña, habrá evidencias de que las diferencias observadas entre los grupos son fruto del azar involucrado al generar los grupos y no del tratamiento.  Si en cambio esta proporcion es suficientemente pequeña, habra evidencias de que las doferencias observadas son causadas por el tratamiento, y no por la configuracion aleatoria de los grupos. Pero hay que delimitar un punto a partir del cual considerar esta proporcion como suficientmiente pequeña o no.

6) Si el p-value calculado en 5) es mayor que cierto limite $\alpha$ , no se podra rechazar H_0 en favor de H_1 puesto que se entiende que no hay suficiente evidencia. En caso de p-value menor que \alpha se acepta H_1 dada la evidencia dispobible. Normalmente se considera \alpha = 0.05 , 0.1 0 o 0.01 , en funcion del nivel de evidencia exigido. A mayor alpha, menos evidencia se exige para aceptar H_1 , cuando menor alpha mayor evidencia exigida para aceptar H_1

En este ejemplo se emplea como estadístico la media pero podría ser cualquier otro (mediana, varianza...)

Para que el test sea exacto deberia tomarse $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$ , que es el numero de diferentes combinaciones  para distribuir $n_t + n_c$  observaciones en dos grupos de tamaños $n_t$ y $n_c$ 

Y ademas las B pares de muestras extraidas deberian de ser distintos (sin que importe el orden de los elementos, es decir dos muestras con los mismos elementos pero en diferente orden son consideradas iguales).

El problema de este enfoque exacto es que para tamaños moderados de $n_t + n_c$ se llegan a valores inasumibles de $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$

Por ejemplo:

In [ ]:
from math import factorial

In [ ]:
factorial(30) / (factorial(20)*factorial(10))

30045015.0

In [ ]:
factorial(30) / (factorial(15)*factorial(15))

155117520.0

La única condición necesaria para un test de permutaciones se conoce como exchangeability, según la cual, todas las posibles permutaciones tienen la misma probabilidad de ocurrir siendo cierta la hipótesis nula.

Los test de permutación son test de significancia y, por lo tanto, se emplean para calcular p-values. Es importante tener en cuenta que, las conclusiones extraídas, solo son aplicables a diseños de tipo experimental. Estudios en los que, tras haber elegido los sujetos, se realiza una asignación aleatoria a los diferentes grupos.

Nota: si bien se utiliza el término permutación, se trata realmente de combinaciones, ya que el orden de las observaciones dentro de cada grupo no importa.

Simulación de Monte Carlo

En los test de permutación, el p-value obtenido es exacto ya que se calculan todas las posibles combinaciones de las observaciones. Esto resulta complicado o imposible cuando el tamaño muestral es mediano o grande. Por ejemplo, existen más de 155 millones combinaciones distintas para distribuir 30 observaciones en dos grupos de 15 cada uno.



(m+n)!m!n!=(15+15)!15!15!=155117520
 


In [ ]:
# Número total de combinaciones
# ==============================================================================
from math import factorial
print(f"Total combinaciones = {factorial(15 + 15) / (factorial(15)*factorial(15))}")



Total combinaciones = 155117520.0
La simulación de Monte Carlo o randomization test consiste en emplear únicamente una muestra aleatoria de todas las posibles combinaciones, evitando así tener que calcularlas todas. Al no calcularse todas las posibles combinaciones, no es un test exacto sino aproximado.

El método de Monte Carlo no es insesgado. Una de las correcciones más empleadas es la propuesta por Davison and Hinkley):

pvalue=r+1k+1
 
siendo r el número de combinaciones igual o más extremas que el estadístico observado y k el número de combinaciones empleadas en la simulación.

A pesar de que la corrección es mínima cuando el número de simulaciones es alto, presenta la ventaja de que, si el valor observado es mayor que cualquiera de los calculados, el p-value es muy bajo pero no cero. Por facilidades de cálculo, se suele emplear un número de permutaciones terminado en 9 (4999, 9999).

¿Cuántas permutaciones (combinaciones) son necesarias?

No hay un número predefinido de combinaciones que asegure buenos resultados. Una forma de guiar la elección es tener en cuenta el valor mínimo capaz de ser detectado. Por ejemplo, con 1000 permutaciones, el menor valor de p-values posible es  pvalue=1+11000+1=0.001998 .

A pesar de ello, la estabilidad y el error de la estimación varían mucho en función de los datos disponibles. Suele encontrarse como recomendación empírica emplear al menos 2000 iteraciones.

¿Cómo interpretar el p-value obtenido?

El p-value es la probabilidad de obtener un valor del estadístico igual o más extremo que el valor observado, si la hipótesis nula es cierta  P(datos|H0) . Es importante no confundir esto con la probabilidad de que la hipótesis nula sea cierta  P(H0|datos) . Si el valor es muy pequeño, se puede interpretar como una evidencia en contra de la hipótesis nula, pero no como una confirmación de las hipótesis alternativas.

## Bibliografia

apuntes Juanmi

https://www.cienciadedatos.net/documentos/pystats03-test-permutacion-python.html